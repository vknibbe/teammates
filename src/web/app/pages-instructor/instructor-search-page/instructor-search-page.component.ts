import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { finalize } from 'rxjs/operators';
import { HttpRequestService } from '../../../services/http-request.service';
import { LoadingBarService } from '../../../services/loading-bar.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { ErrorMessageOutput } from '../../error-message-output';
import { StudentListSectionData } from '../student-list/student-list-section-data';

interface SearchResult {
  searchFeedbackSessionDataTables: SearchFeedbackSessionDataTable[];
  searchStudentsTables: SearchStudentsTable[];
}

interface SearchFeedbackSessionDataTable {
  something: any;
}

/**
 * Data object for communciation with the child student result component
 */
export interface SearchStudentsTable {
  courseId: string;
  sections: StudentListSectionData[];
}

/**
 * Data object for communciation with the child search bar component
 */
export interface SearchQuery {
  searchKey: string;
  searchStudents: boolean;
  searchFeedbackSessionData: boolean;
}

/**
 * Instructor search page.
 */
@Component({
  selector: 'tm-instructor-search-page',
  templateUrl: './instructor-search-page.component.html',
  styleUrls: ['./instructor-search-page.component.scss'],
})
export class InstructorSearchPageComponent implements OnInit {

  searchKey: string = '';
  studentTables: SearchStudentsTable[] = [];
  fbSessionDataTables: SearchFeedbackSessionDataTable[] = [];

  constructor(private route: ActivatedRoute,
              private httpRequestService: HttpRequestService,
              private statusMessageService: StatusMessageService,
              private loadingBarService: LoadingBarService) { }

  ngOnInit(): void {
    this.route.queryParams.subscribe((queryParams: any) => {
      if (queryParams.studentSearchkey) {
        this.searchKey = queryParams.studentSearchkey;
      }
      if (this.searchKey) {
        this.search({
          searchKey: this.searchKey,
          searchStudents: true,
          searchFeedbackSessionData: false,
        });
      }
    });
  }

  /**
   * Searches for students and questions/responses/comments matching the search query.
   */
  search(searchQuery: SearchQuery): void {
    this.loadingBarService.showLoadingBar();
    const paramMap: { [key: string]: string } = {
      searchkey: searchQuery.searchKey,
      searchstudents: searchQuery.searchStudents.toString(),
      searchfeedbacksessiondata: searchQuery.searchFeedbackSessionData.toString(),
    };
    this.httpRequestService.get('/studentsAndSessionData/search', paramMap)
        .pipe(finalize(() => this.loadingBarService.hideLoadingBar()))
        .subscribe((resp: SearchResult) => {
          this.studentTables = resp.searchStudentsTables;
          this.fbSessionDataTables = resp.searchFeedbackSessionDataTables;
          const hasStudents: boolean = !!(this.studentTables && this.studentTables.length);
          const hasFbSessionData: boolean = !!(this.fbSessionDataTables && this.fbSessionDataTables.length);
          if (!hasStudents && !hasFbSessionData) {
            this.statusMessageService.showWarningMessage('No results found.');
          }
        }, (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorMessage(resp.error.message);
        });
  }

}
