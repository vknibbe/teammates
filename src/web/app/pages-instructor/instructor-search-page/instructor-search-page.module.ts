import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { StudentListModule } from '../student-list/student-list.module';
import { InstructorSearchBarComponent } from './instructor-search-bar/instructor-search-bar.component';
import { InstructorSearchPageComponent } from './instructor-search-page.component';
import { StudentResultTableComponent } from './student-result-table/student-result-table.component';

/**
 * Module for instructor search page.
 */
@NgModule({
  declarations: [
    InstructorSearchPageComponent,
    InstructorSearchBarComponent,
    StudentResultTableComponent,
  ],
  exports: [
    InstructorSearchPageComponent,
    InstructorSearchBarComponent,
    StudentResultTableComponent,
  ],
  imports: [
    CommonModule,
    StudentListModule,
    FormsModule,
    RouterModule,
    NgbModule,
  ],
})
export class InstructorSearchPageModule { }
