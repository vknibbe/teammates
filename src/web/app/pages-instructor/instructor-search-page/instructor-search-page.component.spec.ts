import { HttpClientTestingModule } from '@angular/common/http/testing';
import { Component, Input } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBarModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { JoinState } from '../../../types/api-output';
import { InstructorSearchPageComponent, SearchStudentsTable } from './instructor-search-page.component';

@Component({ selector: 'tm-instructor-search-bar', template: '' })
class InstructorSearchBarStubComponent {
  @Input() searchKey: string = '';
}
@Component({ selector: 'tm-student-result-table', template: '' })
class StudentResultTableStubComponent {
  @Input() studentTables: SearchStudentsTable[] = [];
}

describe('InstructorSearchPageComponent', () => {
  let component: InstructorSearchPageComponent;
  let fixture: ComponentFixture<InstructorSearchPageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        InstructorSearchPageComponent,
        InstructorSearchBarStubComponent,
        StudentResultTableStubComponent,
      ],
      imports: [
        HttpClientTestingModule,
        RouterTestingModule,
        MatSnackBarModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorSearchPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a search key', () => {
    component.searchKey = 'TEST';
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a student table', () => {
    component.studentTables = [
      {
        courseId: 'test.tes-demo',
        sections: [
          {
            sectionName: 'Tutorial Group 1',
            isAllowedToViewStudentInSection: true,
            isAllowedToModifyStudent: true,
            students: [
              {
                name: 'Alice Betsy',
                email: 'alice.b.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
              {
                name: 'Benny Charles',
                email: 'benny.c.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
              {
                name: 'Danny Engrid',
                email: 'danny.e.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
              {
                name: 'Emma Farrell',
                email: 'emma.f.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
            ],
          },
        ],
      },
    ];

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
