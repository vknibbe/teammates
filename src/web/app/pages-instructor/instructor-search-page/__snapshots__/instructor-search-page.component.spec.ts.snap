// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorSearchPageComponent should snap with a search key 1`] = `
<tm-instructor-search-page
  fbSessionDataTables={[Function Array]}
  httpRequestService={[Function HttpRequestService]}
  loadingBarService={[Function LoadingBarService]}
  route={[Function ActivatedRoute]}
  searchKey={[Function String]}
  statusMessageService={[Function StatusMessageService]}
  studentTables={[Function Array]}
>
  <tm-instructor-search-bar
    ng-reflect-search-key="TEST"
  />
</tm-instructor-search-page>
`;

exports[`InstructorSearchPageComponent should snap with a student table 1`] = `
<tm-instructor-search-page
  fbSessionDataTables={[Function Array]}
  httpRequestService={[Function HttpRequestService]}
  loadingBarService={[Function LoadingBarService]}
  route={[Function ActivatedRoute]}
  searchKey=""
  statusMessageService={[Function StatusMessageService]}
  studentTables={[Function Array]}
>
  <tm-instructor-search-bar
    ng-reflect-search-key=""
  /><tm-student-result-table
    ng-reflect-student-tables="[object Object]"
  />
</tm-instructor-search-page>
`;

exports[`InstructorSearchPageComponent should snap with default fields 1`] = `
<tm-instructor-search-page
  fbSessionDataTables={[Function Array]}
  httpRequestService={[Function HttpRequestService]}
  loadingBarService={[Function LoadingBarService]}
  route={[Function ActivatedRoute]}
  searchKey=""
  statusMessageService={[Function StatusMessageService]}
  studentTables={[Function Array]}
>
  <tm-instructor-search-bar
    ng-reflect-search-key=""
  />
</tm-instructor-search-page>
`;
